### PR TITLE
CI: undeploy entitlement at the end of the testing

### PR DESCRIPTION
**Pull request by kpouget**
* 755ce46 - CI: undeploy entitlement at the end of the testing

---

* 35e61d5 - gpu_operator_deploy_from_operatorhub: capture the catalog-operator logs when the packagemanifest cannot be found

---

### CODE REVIEW

**Comment by omertuc** [comment]
Does the EXIT trap encompass job interrupt signals (e.g. those caused by a force push)?

**Comment by kpouget** [comment]
> Does the EXIT trap encompass job interrupt signals (e.g. those caused by a force push)?

I'll try to interrupt this testing to get an answer!

**Comment by dagrayvid** [comment]
Looks like a flake...

/test gpu-operator-e2e

**Comment by kpouget** [comment]
cluster failed to deploy ....
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/192/pull-ci-openshift-psap-ci-artifacts-master-gpu-operator-e2e/1404752477768650752

**Comment by kpouget** [comment]
```

2021-06-17 09:39:22,873 p=4078 u=psap-ci-runner n=ansible | roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml:15
2021-06-17 09:39:22,873 p=4078 u=psap-ci-runner n=ansible | TASK: gpu_operator_deploy_from_operatorhub : Wait for the GPU Operator to be available or its catalog to be fully populated
2021-06-17 09:39:22,873 p=4078 u=psap-ci-runner n=ansible | 
2021-06-17 09:39:22,873 p=4078 u=psap-ci-runner n=ansible | <command> oc get packagemanifests/gpu-operator-certified -n openshift-marketplace || test $(oc get -oyaml CatalogSource/certified-operators -n openshift-marketplace '-ojsonpath={.status.connectionState.lastObservedState}{"\n"}') == "READY"
2021-06-17 09:39:22,873 p=4078 u=psap-ci-runner n=ansible | 
2021-06-17 09:39:22,873 p=4078 u=psap-ci-runner n=ansible | <stderr> Error from server (NotFound): packagemanifests.packages.operators.coreos.com "gpu-operator-certified" not found
2021-06-17 09:39:22,883 p=4078 u=psap-ci-runner n=ansible | 
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/192/pull-ci-openshift-psap-ci-artifacts-master-gpu-operator-e2e/1405438613453279232/artifacts/gpu-operator-e2e/presubmit-operatorhub/artifacts/009__gpu-operator__deploy_from_operatorhub/_ansible.log
I think the error comes from there:
```
time="2021-06-17T09:46:56Z" level=warning msg="error encountered while listing bundles: rpc error: code = DeadlineExceeded desc = context deadline exceeded" catalog="{community-operators openshift-marketplace}"
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/192/pull-ci-openshift-psap-ci-artifacts-master-gpu-operator-e2e/1405438613453279232/artifacts/gpu-operator-e2e/gather-extra/artifacts/pods/openshift-operator-lifecycle-manager_catalog-operator-54d6b786b5-llw2p_catalog-operator.log

**Comment by kpouget** [comment]
@omertuc @dagrayvid 
Tests are failing successfully, the "finalizers" are executed as expected:
see [here](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/192/pull-ci-openshift-psap-ci-artifacts-master-sro-e2e-master/1423600059122978816/artifacts/e2e-master/nightly/artifacts/) for SRO: ` 006__entitlement__undeploy`
and [there](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/192/pull-ci-openshift-psap-ci-artifacts-master-gpu-operator-e2e/1423585090780270592/artifacts/gpu-operator-e2e/presubmit-operatorhub/artifacts/) for the GPU Operator: 
- `006__gpu-operator__must-gather`
- `007__entitlement__undeploy`

the tests fail because of https://github.com/openshift-psap/ci-artifacts/pull/233, but at least we know it won't make the CI all 🟢 again ;-)

File: roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml [hunk 1]
```diff
@@ -26,13 +26,23 @@
       > {{ artifact_extra_logs_dir }}/gpu_operator_packagemanifest.json
 
   rescue:
+  - name: Capture the Catalog Operator logs (debug)
```
Comment (omertuc) [inline roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml:29]: Is this related to this PR?

Comment (kpouget) [reply roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml:29]: not related to the title of this PR, but separated into a dedicated commit (https://github.com/openshift-psap/ci-artifacts/pull/192/commits/fbad26a7b1f6c60324204f77c37449e84c2d8ed2) [and I had to update the commits because if got messed up]

I usually ship minor fixes along with the bigger PRs to avoid extra rounds of OCP testing (that's fully incompatible is squash-merging, but the default here is rebase-merging)

**Comment by kpouget** [comment]
/test gpu-operator-e2e

and ping for review @omertuc

**Comment by kpouget** [comment]
@omertuc can this be merged?

File: roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml [hunk 2]
```diff
@@ -26,13 +26,23 @@
...
+    shell:
+      oc logs deployment.apps/catalog-operator
```
Comment (omertuc) [inline roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml:31]: Why do we even collect these logs, aren't they available in `gather-extra` anyway?

File: build/root/usr/local/bin/ci_entrypoint_gpu-operator.sh [hunk 3]
```diff
@@ -21,31 +23,35 @@ prepare_cluster_for_gpu_operator() {
 }
 
 collect_must_gather() {
-    set +x
-    echo "Running gpu-operator_gather ..."
-    /usr/bin/gpu-operator_gather &> /dev/null
+    run_in_sub_shell() {
```
Comment (omertuc) [inline build/root/usr/local/bin/ci_entrypoint_gpu-operator.sh:26]: Function name is a bit misleading, it sounds like this function will run something in a sub shall, rather than be ran in a sub shell itself

Comment (omertuc) [reply build/root/usr/local/bin/ci_entrypoint_gpu-operator.sh:26]: Maybe rename it to `collect` or something

File: roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml (hunk 2 above)
Comment (kpouget) [reply roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml:31]: they are ... as long as we're running in Prow CI

and from my point of view/experience, it's the most efficient way to troubleshot when then likely cause of outage are stored along with the test results. The `(debug)` tag in the name informally says that this task isn't part of the test but only there for debugging/troubleshooting/observability

Comment (omertuc) [reply roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml:31]: > they are ... as long as we're running in Prow CI
Fair enough

> it's the most efficient way to troubleshot when then likely cause of outage are stored along with the test results

With the existence of things like https://github.com/kxr/o-must-gather, I must say I disagree - it's easier for me to find things with `omg` on the Prow must-gather than having to "know" about specialized artifacts that we create

File: build/root/usr/local/bin/ci_entrypoint_gpu-operator.sh (hunk 3 above)
Comment (kpouget) [reply build/root/usr/local/bin/ci_entrypoint_gpu-operator.sh:26]: don't know, it's a local function witht the subshell invocation a few lines after, I don't find it hard to understand ...
```
run_in_sub_shell() {
...
}
    # run the function above in a subshell to avoid polluting the local `env`.
    typeset -fx run_in_sub_shell
    bash -c run_in_sub_shell
```
that's similar to 
```
cb() {
 ... do callback ...
}
iterate_over_something(cb)
```

Comment (omertuc) [reply build/root/usr/local/bin/ci_entrypoint_gpu-operator.sh:26]: I personally was mislead / confused by it, so I thought I should explain why. Doesn't really matter

File: roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml (hunk 2 above)
Comment (kpouget) [reply roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml:31]: if the test fail, how do you (an external observer) know where to go for troubleshooting the issue?

as the role implementor, I give you the most likely culprint

ideally (my ideal at least), for every known `flake`, you wouldn't have to go to `gather-extra` to know what issue caused the flake

Comment (omertuc) [reply roles/gpu_operator_deploy_from_operatorhub/tasks/deploy_from_catalog.yml:31]: Yeah that's a good point
